sublime-merge: init at 1107

zookatron opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Commit:
sublime-merge: init at 1107

worldofpeace commented on pkgs/applications/version-management/sublime-merge/common.nix:
```diff
@@ -0,0 +1,107 @@
+      runHook preInstall
+
+      # Correct sublime_merge.desktop to exec 'sublime_merge' instead of /opt/sublime_merge
+      substituteInPlace "sublime_merge.desktop" --replace "/opt/sublime_merge/sublime_merge" "$out/sublime_merge"
```
Why fix this here and then again below?

zookatron replied:
I guess there isn't really a good reason. I've removed this line and changed the substitution below as suggested.

My thinking was to try to leave the bin derivation in as correct of a state as possible, without any incorrect paths, even though it is only used by the wrapper derivation. But I guess the wrapper derivation will always have to do another substitution anyway so it doesn't really matter.

worldofpeace commented on pkgs/applications/version-management/sublime-merge/common.nix:
```diff
@@ -0,0 +1,107 @@
+    makeWrapper "$sublime_merge/sublime_merge" "$out/bin/sublime_merge";
+    ln -s "$out/bin/sublime_merge" "$out/bin/smerge";
+    mkdir -p "$out/share/applications"
+    substitute "$sublime_merge/sublime_merge.desktop" "$out/share/applications/sublime_merge.desktop" --replace "$sublime_merge/sublime_merge" "$out/bin/sublime_merge"
```
```suggestion
    substitute "$sublime_merge/sublime_merge.desktop" "$out/share/applications/sublime_merge.desktop" --replace "/opt/sublime_merge/sublime_merge" "$out/bin/sublime_merge"
```

tadeokondrak commented on pkgs/applications/version-management/sublime-merge/common.nix:
```diff
@@ -0,0 +1,107 @@
+    '';
+  };
+in stdenv.mkDerivation {
+  name = "sublime-merge-${buildVersion}";
```
```suggestion
  pname = "sublime-merge";
```

zookatron replied:
Hi @tadeokondrak, thanks for the feedback. I was having trouble finding documentation on exactly when to use `pname` vs `name`. Am I correct in understanding that you should use `pname` when you have a separate `version` attribute and `name` when the version is part of the name?

worldofpeace replied:
Yeah, we need documentation on this really soon.

Simply put, the `name` attribute is the derivation name. And when attributes `pname` and `version` are present they get concatenated to `name = "${pname}-${version}"`

> https://github.com/NixOS/nixpkgs/blob/c40f2114992f7387013a6113ebfd99ae63ec4f91/pkgs/stdenv/generic/make-derivation.nix#L181

This exists because you frequently want to reuse the value of `pname` and `version` throughout the expression.

worldofpeace commented on pkgs/applications/version-management/sublime-merge/common.nix:
```diff
@@ -0,0 +1,107 @@
+    ln -s "$out/bin/sublime_merge" "$out/bin/smerge";
+    mkdir -p "$out/share/applications"
+    substitute "$sublime_merge/sublime_merge.desktop" "$out/share/applications/sublime_merge.desktop" --replace "$sublime_merge/sublime_merge" "$out/bin/sublime_merge"
+    for iconSize in $(ls "$sublime_merge/Icon"); do
```
I tend to not use `ls` in scripts ever.

Using substring removal and a glob would be better.
```
for directory in $sublime_merge/Icon/*; do
  size=''${directory##*/}
  mkdir -p "$out/share/icons/hicolor/$size/apps";
  cp "$sublime_merge/Icon/$size/sublime-merge.png" "$out/share/icons/hicolor/$size/apps";
done
```

##### we need the `''` to escape the `$` btw

zookatron replied:
Thanks for the feedback, I've fixed this.

Can you help me understand why it is good to avoid using ls? Doesn't it have reliable behavior, especially in a consistent environment like stdenv?

By the way I ended up using `$(dirname $directory)` instead of `''${directory##*/}` because I am more familiar with it. Could you point me to some documentation where I can learn more about the `##*/` shell construct? I'm having trouble finding anything that describes how that works.

worldofpeace replied:
> By the way I ended up using $(dirname $directory) instead of ''${directory##*/} because I am more familiar with it. Could you point me to some documentation where I can learn more about the ##*/ shell construct? I'm having trouble finding anything that describes how that works.

This should be very helpful https://wiki-dev.bash-hackers.org/syntax/pe#substring_removal

> Can you help me understand why it is good to avoid using ls? Doesn't it have reliable behavior, especially in a consistent environment like stdenv?

The problem is that the output of `ls` is meant to be human readable and so its not meant to parsed.
We have globbing for this purpose, to allow the shell to expand patterns into a list of matching filenames.

Here's a useful resource on that topic
* http://mywiki.wooledge.org/ParsingLs

zookatron replied:
That makes sense, thanks for the links!

worldofpeace reviewed (changes requested):
Looks good, left a few comments.

Also you should add yourself to the maintainers list in a separate commit like
```
maintainers: add zookatron
```
and before the package init

Commit:
maintainers: add zookatron

zookatron commented:
Hi @tadeokondrak @worldofpeace ! I've applied all the requested changes in the recent updates to my branch. Let me know if there's anything else that can be improved!

worldofpeace commented on pkgs/applications/version-management/sublime-merge/common.nix:
```diff
@@ -0,0 +1,105 @@
+    for directory in $sublime_merge/Icon/*; do
+      size=$(basename $directory)
+      mkdir -p "$out/share/icons/hicolor/$size/apps"
+      cp "$sublime_merge/Icon/$size/sublime-merge.png" "$out/share/icons/hicolor/$size/apps"
```
Would it be better to use symbolic links?

zookatron replied:
Yeah, I guess it would make sense to prefer symlinks over copying files wherever possible. I actually mostly copied that particular section from [the Sublime Text 2 derivation](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/sublime/2/default.nix) which I thought was pretty much the standard way of doing things but I can see now that other packages often use symlinks instead for icons. I've fixed this issue in my most recent update to my branch.

worldofpeace reviewed (approved):
Everything builds and executes fine.

worldofpeace commented:
ping @jtojnar @Mic92 because you reviewed https://github.com/NixOS/nixpkgs/pull/47378

jtojnar commented:
It might make sense to change the `sublime3`’s `common.nix` to make it reusable. After all they both use the same base.

zookatron commented:
> It might make sense to change the `sublime3`’s `common.nix` to make it reusable. After all they both use the same base.

Possibly, I don't have enough experience to comment on how nixos commonly does that, but it seems to be that there's enough differences between the sublime3 derivation and the sublime-merge derivation to not really make it worth it. They definitely share a few important bits of logic like the `pkexec` fixes, but sublime-merge is missing a lot of things that sublime3 does like supporting 32 bit systems, wrapping the `plugin_host`, creating a custom `sublime_bash` and fixing the Packages to use it, etc. Seems to me like you would need enough conditionals and overrides to account for the differences that it would just make the packages more difficult to maintain properly overall, but I don't know.

zookatron commented:
@jtojnar Are there any blockers to having this merged? Would you like me to work on creating the shared `common.nix` file before we commit this? I'm happy to help if that's what you think needs to happen.

jtojnar commented:
Factoring out Sublime packages into a common expression can wait to future pull requests. However, I would like to see all the non-Merge-specific deviations from the Text expression to be applied there as well, in order to make the future unification easier. Basically, we want a minimal `git diff --no-index pkgs/applications/editors/sublime/3/common.nix pkgs/applications/version-management/sublime-merge/common.nix`.

Commit:
sublime-{merge,text}: make common.nix files have a similar structure

In the future it's desired to unify Sublime packages expressions.
So them having a shared structure will make this more achievable.

zookatron commented:
> Factoring out Sublime packages into a common expression can wait to future pull requests. However, I would like to see all the non-Merge-specific deviations from the Text expression to be applied there as well, in order to make the future unification easier. Basically, we want a minimal `git diff --no-index pkgs/applications/editors/sublime/3/common.nix pkgs/applications/version-management/sublime-merge/common.nix`.

@jtojnar I've added a commit that updates the sublime3 and sublime-merge common.nix files to better reflect their shared structure. If you run `git diff --no-index pkgs/applications/editors/sublime/3/common.nix pkgs/applications/version-management/sublime-merge/common.nix` it now has a minimal set of diffs that more accurately reflects the necessary build differences between the two packages. Incidentally this also fixes a bug in sublime3 where the icon files where not being linked to the proper directory preventing the desktop environment from displaying their custom icons.

worldofpeace commented:
Sorry that I didn't merge that for you sooner @zookatron 

Thank you for your contribution and co-operation :sparkling_heart:

zookatron commented:
No problem, thanks @worldofpeace!
